### ISSUE TITLE
ObjC method that returns AnyPromise is not treated as AnyPromise in Swift

### ISSUE THREAD

**Issue by dbachrach** [mxcl/PromiseKit#225]
We are porting our PromiseKit 1 based library to PromiseKit 2. Our library is entirely ObjC, and my goal is to allow consumption of the library in ObjC or swift. I've changed all of our methods to return AnyPromise instead of PMKPromise.

```objc
+ (AnyPromise*)fetch;
```

When I look at the generated swift header for our method I see this:

```swift
class func fetch() -> PMKPromise!
```

So when we try to call `fetch` from Swift, it's not very pretty:

```swift
let p = MHObject.fetch()
```

P is of type `PMKPromise!`, when I really wanted it to be of type `AnyPromise`. So I now have to write this instead:

```swift
let p = MHObject.fetch() as! AnyPromise
```

This breaks easy chaining because I have to `as!` all my promises. Is there a way to perfectly bridge ObjC `AnyPromise` into Swift `AnyPromise`?

**Comment by mxcl** [comment]
AnyPromise is a type alias of PMKPromise. So you shouldn't need to cast. Do you have to?

**Comment by dbachrach** [comment]
Right. That's what I thought too, but swift is complaining. I'm on Xcode 6.4. Here is the swift code that works:

```swift
let p = MHObject.fetch() as! AnyPromise
p.then { obj -> Void in
    // do something
}
```

If I remove the `as!` like this:

```swift
let p = MHObject.fetch()
p.then { obj -> Void in
    // do something
}
```

I get the following error:

```
ViewController.swift:28:12: Cannot invoke 'then' with an argument list of type '((_) -> Void)'
````

**Comment by dbachrach** [comment]
2.1.1 PromiseKit as well.

**Comment by mxcl** [comment]
I think this is just a Swift compile error. You must specialize obj. Look at the signature for then, it's generic. Compare with the 2.0 release notes.

**Comment by mxcl** [comment]
```swift
let p = MHObject.fetch()
p.then { (obj: AnyObject) -> Void in
    // do something
}
```

I agree this sucks. Maybe then should just push `AnyObject?` from an AnyPromise.

**Comment by dbachrach** [comment]
```swift
let p = MHObject.fetch()
p.then { (obj: AnyObject?) -> Void in
    // do nothing
}
```
Doing the above I still get the error: `Cannot invoke 'then' with an argument list of type '((AnyObject?) -> Void)'`

Simply by `as!`ing to an AnyPromise, the code then works. So it seems like somehow the compiler doesn't think that PMKPromise and AnyPromise are quite the same thing. When I do the `as!` I get this warning:

```
Cast from 'PMKPromise!' to unrelated type 'AnyPromise' always fails
````

**Comment by mxcl** [comment]
Can you provide a full header of one these?

**Comment by dbachrach** [comment]
I sent you an email with our project. Thanks for the help!

**Comment by mxcl** [comment]
I appreciate the full test case, I'll make sure to figure out a fix for you!

**Comment by mxcl** [comment]
OK, I understand what's the problem more or less.

It’s a module problem. Your library CoreHound imports PromiseKit, but the module system gets confused between the objc and Swift components and ends up only importing the ObjC side of PromiseKit. You can verify this by trying to complete off of `p2`: it provides completions for ObjC-ish functions (note that they have `Void` parameters).

This is quite a substantial problem and will require a good deal of work for me to fix it.

I tried initially to replace all instances of `#import <PromiseKit/PromiseKit.h> in public headers in CoreHound and other pods with `@class PMKPromise` and then moving the #import to the `.m` files.

This worked, but then for some reason `ViewController.swift` still wouldn't compile as no function that returned `PMKPromise` had public access.

I'll have to give this some thought.

TL;DR: Any module that exposes an Objective-C interface that returns promises and which is then imported by Swift code will fail.

**Comment by mxcl** [comment]
In fact, it imports almost all the Swift stuff, just not the AnyPromise part of the Swift stuff. Weird.

**Comment by mxcl** [comment]
I'm still trying to solve this. Apologies for the delay.

**Comment by dbachrach** [comment]
No worries. Thanks for taking a look at it. Let me know if you need anything from me.

**Comment by mxcl** [comment]
Have a solution. It’s deep module-foo. Dunno what specifically the problem was, but it is fixed.

**Comment by dbachrach** [comment]
@mxcl Mad work! Thanks. Will try this out this weekend.
